### ISSUE TITLE
Add analyzer and documentation to migrate customers from MonoPInvokeCallbackAttribute to UnmanagedCallersOnlyAttribute

### ISSUE THREAD

**Issue by CoffeeFlux** [dotnet/runtime#47177]
We are recommending use of `UnmanagedCallersOnlyAttribute` moving forward, but its usage isn't quite the same. We should have an analyzer to detect use of `MonoPInvokeCallbackAttribute` and documentation on how to migrate over where appropriate.

cc: @marek-safar

**Comment by lambdageek** [comment]
There are no plans at this time to drop support for `MonoPInvokeCallbackAttribute` from MonoVM. That is, if you declare your own `MonoPInvokeCallbackAttribute` or use [`ObjCRuntime.MonoPInvokeCallbackAttribute`](https://docs.microsoft.com/en-us/dotnet/api/objcruntime.monopinvokecallbackattribute?view=xamarin-ios-sdk-12) it will continue to work. But `UnmanagedCallersOnlyAttribute` is preferred.

It's not quite the same however since it doesn't use a delegate for the method signature, and it also makes the method uncallable from managed.

/cc @rolfbjarne @AaronRobinsonMSFT

**Comment by AaronRobinsonMSFT** [comment]
>It's not quite the same however since it doesn't use a delegate for the method signature, and it also makes the method uncallable from managed.

Yep. There is a lot to consider if one goes with `UnmanagedCallersOnly` - basically all types must be blittable. I don't see this as a being particularly good idea unless the consumers want to go down the path of manually marshalling all non-blittable types - that seems like a complicated message to suggest. It would become:

"Consider `UnmanagedCallersOnly` since all arguments of blittable"

or

"Converting all arguments to blittable will permit usage of `UnmanagedCallersOnly`"

or the most scary

"Read up on blittability and convert to blittables types to use `UnmanagedCallersOnly`"

What sort of guidance is being considered here?

**Comment by marek-safar** [comment]
What are the benefits of using UnmanagedCallersOnlyAttribute instead of MonoPInvokeCallbackAttribute ?

**Comment by rolfbjarne** [comment]
One reason I can think of is that you get a nice compiler error if you use UnmanagedCallersOnlyAttribute wrong.

On the other hand, if the signature of the delegate passed to MonoPInvokeCallbackAttribute doesn't match the signature of the method, the AOT compiler will abort with a very unhelpful error.

I also believe that the error message you get if you add MonoPInvokeCallbackAttribute to an instance method is less than ideal.

**Comment by CoffeeFlux** [comment]
You also have the ability to set the `EntryPoint`, which I believe is something @jonpryor had asked for with `MonoPInvokeAttribute`.

**Comment by lambdageek** [comment]
For library authors [UnmanagedCallersOnly] is nice because you can use it on all .NET5 runtimes.

The different marshalling behavior does mean that it might be a big ask to migrate all users over.  So maybe this is too ambitious until we have a source generator that can make marshalling wrappers on reverse pinvokes.

**Comment by marek-safar** [comment]
I think this is nice to have for 6 but we should replace the usage of this attribute in SDKs to not encourage further usage of the attribute.

**Comment by CoffeeFlux** [comment]
It looks like rolf opened an issue for iOS, and I opened https://github.com/dotnet/runtime/issues/47176 for our samples in dotnet/runtime, so I guess that just leaves Android.
